File: Problems/02-Symmetric-Tree/Symmetric_Tree.cpp
#include <bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	bool check(TreeNode* p, TreeNode* q)
	{
		if (p == nullptr and q == nullptr) return true;
		if (p == nullptr or q == nullptr) return false;
		if (p->val != q->val) return false;
		return check(p->left, q->right) && check(p->right, q->left);
	}
    bool isSymmetric(TreeNode* root) {
        return check(root, root);
    }
};

File: Problems/02-Symmetric-Tree/symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root) -> bool:
        return self.isMirror(root,root)
    
    def isMirror(self,p1,p2) ->bool:
        if p1 is None and p2 is None:
            return True
        if p1 and p2 and p1.val == p2.val:
            return self.isMirror(p1.left,p2.right) and self.isMirror(p1.right,p2.left)
        else:
            return False

    